### PR TITLE
linux_sunlight: init at 6.1.21

### PR DESCRIPTION
**Pull request by ionutnechita**
Description:
 - linux_sunlight: init at 6.1.21

Signed-off-by: Ionut Nechita <ionut_n2001@yahoo.com>

###### Description of changes

linux_sunlight: init at 6.1.21

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by Atemu — COMMENTED** [review]
`sunlight-kernel: 6.1.15 lts` -> `sunlight-kernel: init at 6.1.15`

I can't really review this. I don't know the kernel nor what it's useful for.

File: pkgs/os-specific/linux/kernel/sunlight-kernels.nix [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    extraMeta = {
+      branch = lib.versions.majorMinor version;
+      maintainers = with lib.maintainers; [ fortuneteller2k lovesegfault atemu ];
+      description = "Sunlight Kernel. Built with custom settings and new features built to provide a stable, responsive and smooth desktop experience";
+      broken = stdenv.isAarch64;
+    };
+  } // (args.argsOverride or { }));
+in
+{
+  lts = sunlightKernelFor ltsVariant;
```
[the remark this answers is not in the document]
Comment (ionutnechita) [reply pkgs/os-specific/linux/kernel/sunlight-kernels.nix:68]: Hi @Atemu,

Yes, i working on main variant.
Now exist lts variant from stable version.

I am the author of this kernel.
This kernel is optimized for desktop responsive and performance improvement with several changes on scheduler.

[the remark this answers is not in the document]
Comment (ionutnechita) [reply pkgs/os-specific/linux/kernel/sunlight-kernels.nix:68]: Improve for AMD cpu's, BORE scheduler, KVM and i915.

File: pkgs/os-specific/linux/kernel/sunlight-kernels.nix [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      HZ_1000 = no;
+
+      SCHEDSTATS = lib.mkOverride 60 yes;
+    };
+
+    extraMeta = {
+      branch = lib.versions.majorMinor version;
+      maintainers = with lib.maintainers; [ fortuneteller2k lovesegfault atemu ];
+      description = "Sunlight Kernel. Built with custom settings and new features built to provide a stable, responsive and smooth desktop experience";
+      broken = stdenv.isAarch64;
```
Comment (lovesegfault) [inline pkgs/os-specific/linux/kernel/sunlight-kernels.nix:63]: Is it just aarch64 or anything other than `x86_64-linux`?

File: pkgs/os-specific/linux/kernel/sunlight-kernels.nix [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      WINESYNC = module;
+
+      # Preemptive Full Tickless Kernel at 833Hz
+      LATENCYTOP = yes;
+
+      PREEMPT = lib.mkOverride 60 yes;
+      PREEMPT_VOLUNTARY = lib.mkForce no;
+      PREEMPT_NONE = lib.mkForce no;
+
+      HZ = freeform "833";
```
Comment (lovesegfault) [inline pkgs/os-specific/linux/kernel/sunlight-kernels.nix:52]: This is oddly specific, where does this value come from?

Comment (ionutnechita) [reply pkgs/os-specific/linux/kernel/sunlight-kernels.nix:52]: 833 Hz is alternative to 1000 Hz. I choose this value for a balance between latency, performance and low power consumption.

Comment (Atemu) [reply pkgs/os-specific/linux/kernel/sunlight-kernels.nix:52]: Ideally explain such changes in a comment right in the code.

Comment (ionutnechita) [reply pkgs/os-specific/linux/kernel/sunlight-kernels.nix:55]: We brought explain with comment right in the code.

File: pkgs/top-level/linux-kernels.nix [hunk 4]
```diff
@@ -241,6 +241,12 @@ in {
     linux_xanmod_stable = xanmodKernels.main;
     linux_xanmod_latest = xanmodKernels.main;
 
+    sunlightKernels = callPackage ../os-specific/linux/kernel/sunlight-kernels.nix {
+      kernelPatches = [];
```
Comment (Atemu) [inline pkgs/top-level/linux-kernels.nix:245]: Do you incorporate our patches?

Comment (ionutnechita) [reply pkgs/top-level/linux-kernels.nix:245]: No, i not have patch files.

You suggest to remove this?

Comment (Atemu) [reply pkgs/top-level/linux-kernels.nix:245]: In that case you should probably put the kernel patches in that we include in our other kernels of this version.

**Comment by Atemu** [comment]
No merge commits, please just rebase.

File: pkgs/top-level/linux-kernels.nix (hunk 4 above)
Comment (Atemu) [reply pkgs/top-level/linux-kernels.nix]: I mean the kernels we put into every kernel in Nixpkgs. Please take a look at how it's done for the zen kernels or xanmod.

Comment (ionutnechita) [reply pkgs/top-level/linux-kernels.nix]: I checked how it looks in xanmod and zen, in the patch area(kernelPatches), I made an update for this aspect.

**Comment by ionutnechita** [comment]
Hi @Atemu,

At the moment this commit works, I checked it on my laptop and I can't find any problem. you can do a review to go and have it as a package.

File: pkgs/os-specific/linux/kernel/sunlight-kernels.nix [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      HZ_833 = yes;
+      HZ_1000 = no;
+
+      SCHEDSTATS = lib.mkOverride 60 yes;
+    };
+
+    extraMeta = {
+      branch = lib.versions.majorMinor version;
+      maintainers = with lib.maintainers; [ fortuneteller2k lovesegfault atemu ];
```
Comment (Atemu) [inline pkgs/os-specific/linux/kernel/sunlight-kernels.nix:62]: Please remove us xanmod maintainers and add yourself.

**Review by Atemu — CHANGES_REQUESTED** [review]
I can do a proper review no sooner than this Sunday.

File: pkgs/os-specific/linux/kernel/sunlight-kernels.nix [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+      HZ = freeform "833";
+      HZ_833 = yes;
+      HZ_1000 = no;
+
+      SCHEDSTATS = lib.mkOverride 60 yes;
+    };
+
+    extraMeta = {
+      branch = lib.versions.majorMinor version;
+      maintainers = with lib.maintainers; [ ionutnechita ];
```
Comment (ionutnechita) [inline pkgs/os-specific/linux/kernel/sunlight-kernels.nix:62]: I added on me for maintainers.

File: pkgs/os-specific/linux/kernel/sunlight-kernels.nix (hunk 5 above)
Comment (ionutnechita) [reply pkgs/os-specific/linux/kernel/sunlight-kernels.nix]: This has been resolved.

File: pkgs/os-specific/linux/kernel/sunlight-kernels.nix (hunk 6 above)
Comment (ionutnechita) [reply pkgs/os-specific/linux/kernel/sunlight-kernels.nix:62]: I have one fail from ofborg-eval: 

nix-env failed:
error: undefined variable 'ionutnechita'

Comment (ionutnechita) [reply pkgs/os-specific/linux/kernel/sunlight-kernels.nix:62]: in this case is necessary to add my name on maintainer-list.nix

File: pkgs/os-specific/linux/kernel/sunlight-kernels.nix [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+      # 833 Hz is alternative to 1000 Hz.
+      # Selected value for a balance between latency, performance and low power consumption.
+      HZ = freeform "833";
+      HZ_833 = yes;
+      HZ_1000 = no;
+
+      SCHEDSTATS = lib.mkOverride 60 yes;
+    };
+
+    extraMeta = {
```
Comment (davidak) [inline pkgs/os-specific/linux/kernel/sunlight-kernels.nix:60]: can you add `homepage`?

Comment (ionutnechita) [reply pkgs/os-specific/linux/kernel/sunlight-kernels.nix:60]: I don't have a homepage or website at the moment. I am also working on this aspect.

**Review by Atemu — CHANGES_REQUESTED** [review]
Please split your commit into 3.

As a rule of thumb, if you need bullet points in your commit message, you should probably make each a separate commit.

File: pkgs/os-specific/linux/kernel/common-config.nix [hunk 8]
```diff
@@ -966,7 +966,7 @@ let
       FSL_MC_UAPI_SUPPORT = mkIf (stdenv.hostPlatform.system == "aarch64-linux") (whenAtLeast "5.12" yes);
 
       ASHMEM =                 { optional = true; tristate = whenBetween "5.0" "5.18" "y";};
-      ANDROID =                { optional = true; tristate = whenAtLeast "5.0" "y";};
+      ANDROID =                { optional = true; tristate = whenBetween "5.0" "5.19" "y";};
```
[the remark this answers is not in the document]
Comment (ionutnechita) [reply pkgs/os-specific/linux/kernel/common-config.nix:969]: ANDROID option is not available with 6.0 kernel version.

[the remark this answers is not in the document]
Comment (Atemu) [reply pkgs/os-specific/linux/kernel/common-config.nix:969]: That should be its own PR probably; small and simple. Target it at staging as it's just a warning fix and we don't need that to cause rebuilds.

[the remark this answers is not in the document]
Comment (ionutnechita) [reply pkgs/os-specific/linux/kernel/common-config.nix:969]: Ok, will try to create new PR for isolate this change.

[the remark this answers is not in the document]
Comment (ionutnechita) [reply pkgs/os-specific/linux/kernel/common-config.nix:969]: I created new PR for android option:

https://github.com/NixOS/nixpkgs/pull/223694

File: pkgs/os-specific/linux/kernel/sunlight-kernels.nix (hunk 7 above)
Comment (davidak) [reply pkgs/os-specific/linux/kernel/sunlight-kernels.nix:60]: if you have a readme in the repo that explains what it is etc, that is also fine

Comment (ionutnechita) [reply pkgs/os-specific/linux/kernel/sunlight-kernels.nix:60]: I will try to add more information about it along the way.

**Comment by ionutnechita** [comment]
Hi @Atemu,

Is it possible merge this PR?

**Review by Atemu — CHANGES_REQUESTED** [review]
`sunlight-kernel: maintainers: added new user in list` -> `maintainers: add ionutnechita`

Sign-offs and change-ids aren't necessary here but you can leave them in if you want.

Please squash the two commits which add the sunlight kernel. It's one logical step.

File: pkgs/os-specific/linux/kernel/sunlight-kernels.nix (hunk 2 above)
Comment (ionutnechita) [reply pkgs/os-specific/linux/kernel/sunlight-kernels.nix:64]: For now it's just aarch64. But in the future I will no longer want to be limited to architecture.

**Review by Atemu — DISMISSED** [review]
`maintainers: added new user in list` -> `maintainers: add ionutnechita`

`sunlight-kernel: init at 6.1.21` -> `linux_sunlight: init at 6.1.21`

Please read https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md.

**Comment by Atemu** [comment]
Alright, from the technical side, I have no further objections.

I am however unsure whether we want this kernel in Nixpkgs. I appreciate OSS projects getting themselves packaged but I do not yet see a "real" project here. There is no project page, no community around the project (AFAICT), no reputation, the sunlight kernel is packaged in no other repository I know of and I haven't heard of it before either. Those facts alone don't disqualify it from being a project but they do make me question whether it should be included in an "official" repository such as Nixpkgs.

I personally think the sunlight kernel would be better fit in a user repo such as the AUR or our [NUR](https://github.com/nix-community/NUR) until it has matured a bit. 

Do not get me wrong, I am not opposed to its inclusion and there is no policy against it that I know of, I just don't feel like it really belongs to an official repo just yet.

If someone else makes the call to include it, I will not object but I do not approve either.

**Comment by ionutnechita** [comment]
I understand, I will continue working on the NUR to be able to make a final version.

**Pull request closed without merging**